sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -526,6 +539,18 @@ class Analyzer(
       }
     }
 
+    private def containsUnresolvedNewInstance(expr: Expression): Boolean = expr.find {
+      case n: NewInstance if needOuterScope(n) => true
+      case _ => false
+    }.isDefined
+
+    private def needOuterScope(n: NewInstance): Boolean = {
+      // If this is an inner class of another class, register the outer object in `OuterScopes`.
+      // Note that static inner classes (e.g., inner classes within Scala objects) don't need
+      // outer pointer registration.
+      n.outerPointer.isEmpty && n.cls.isMemberClass && !Modifier.isStatic(n.cls.getModifiers)
```

Comment:
Could this go into `NewInstance`?  it seems like `resolved` should be `false` until this is satisfied.